### PR TITLE
Fix: Serialize resource type for unpersisted records (blank id)

### LINKED ISSUE #2132: unable to serialize unpersisted records
##### expected:
```ruby
ActiveModelSerializer::SerializableResource.new(FactoryGirl.build(:some_object).as_json


"{ 
  "data": {
    "type": "some-objects",
    "id": null,                            #or just not there entirely
    "attributes": { /// }
  }
}"           
```        
##### actual:
```ruby
ActiveModelSerializer::SerializableResource.new(FactoryGirl.build(:some_object).as_json


"{ "data": null }"
```     

#### Steps to reproduce
serialize an activerecord object that doesn't have an id yet
in 0.10.5 it behaves as expected
in 0.10.6 it returns data null

#### Environment

ActiveModelSerializers Version: 0.10.6

Output of `ruby -e "puts RUBY_DESCRIPTION"`: 
ruby 2.3.1p112 (2016-04-26 revision 54768) [x86_64-darwin16]

OS Type & Version: osx 10.12.4

Integrated application and version: rails 4.2

#### Additonal helpful information
This functionality was introduced in #2119 specifically [here](https://github.com/rails-api/active_model_serializers/pull/2119/files#diff-65abd6a4fd7fd55fe487ff138e0841dbR40)
and someone else already complained about it in #2126

I also use this functionality to test my controller endpoints, serializing non-saved objects in order to test POST and PATCH requests.

Why was this functionality added, and is there another way to serialize non-persisted records via json-api adapter?

### LINKED ISSUE #2180: Does not allow serialization for purposes of Creating Resources
#### Expected behavior vs actual behavior
I expect to be able to serialize an object without an id for the purpose creating a resource.
Current behavior returns: `{data: null}`

#### Steps to reproduce
```
let(:user_attributes) { FactoryGirl.build(:user) }
post "/v1/users", params: ActiveModelSerializers::Adapter::JsonApi.new(UserSerializer.new(user_attributes)).serializable_hash
```

#### Environment

ActiveModelSerializers Version *(commit ref if not on tag)*: 
Everything south of `73eae19b3d0b8a3c03fda566a208eeee6720beed`

#### Additonal helpful information
*(e.g., Gemfile.lock, configurations, PR containing a failing test, git bisect results)*

From the spec:
[Creating Resources](http://jsonapi.org/format/#crud-creating)
```A resource can be created by sending a POST request to a URL that represents a collection of resources. The request MUST include a single resource object as primary data. The resource object MUST contain at least a type member.```
This is implying that the resource object MAY contain an id, not MUST.

Also here:
https://github.com/rails-api/active_model_serializers/issues/2132

### PR DESCRIPTION
**Pull request by bf4**
Current behavior when `id` is nil returns: `{data: null}`

Closes #2132
Closes #2180

Regression from v0.10.5 to v0.10.6

Incidental changes in the PR

- genereralized inflection test helper
- refactored resource identifier tests to use more explicit assertions
- rolled resource identifier type tests into jsonapi type tests
- Added regression spec for defined type not being inflected, as referenced in #2211

### CODE REVIEW

**Comment by bf4** [comment]
cc interested parties @InteNs @morhekil @yoda @beet @xn @Genkilabs @gambala

**Comment by xn** [comment]
Thanks!

**Comment by InteNs** [comment]
looks good!

**Comment by bf4** [comment]
Anything else before I release a 0.10.7? I think the only other regression is https://github.com/rails-api/active_model_serializers/issues/2178

**Comment by Genkilabs** [comment]
Terrific! Thank you for this, and for including the CC in the related PR.
^_^

**Comment by beet** [comment]
Awesome, thanks @bf4 !

File: lib/active_model_serializers/adapter/json_api/resource_identifier.rb [hunk 1]
```diff
@@ -38,8 +39,11 @@ def initialize(serializer, options)
         end
 
         def as_json
```
Comment (bf4) [inline lib/active_model_serializers/adapter/json_api/resource_identifier.rb]: arguably, if `serializer.object` is nil, the return value should be `nil`, not `{ type: type}`, and consider it a 'nil' identifier

**Pull request merged**